feat: health check in worker

SantiagoPittella opened the pull request:
This PR adds:
- Health Check endpoints to the remote prover using the `tonic-health` crate, which adds the required endpoints without the need of defining the methods in a `.proto` file.
- Add a background service in the Proxy that periodically checks the health of each worker and removes the ones that are not working.

igamigo commented on bin/tx-prover/src/proxy/mod.rs:
```diff
@@ -239,6 +254,59 @@ impl LoadBalancer {
+    ///
+    /// Performs a health check on each worker using the gRPC health check protocol. If a worker
+    /// is not healthy, it won't be included in the list of healthy workers.
+    async fn check_workers_health(
+        &self,
+        workers: impl Iterator<Item = &mut Worker>,
+    ) -> Vec<Worker> {
```
One thing that we probably want to check in the future (on another PR) is that if a worker does not respond to one health check, it does not necessarily mean it is permanently down (could be related to network transient failures, etc.), so we probably want to have some sort of retries before permanently taking them down.

SantiagoPittella replied:
I opened #1009 to address this.

igamigo reviewed (commented):
Nice! Leaving some comments for now, will take a deeper look (and maybe test it as well) later

bobbinth commented on bin/tx-prover/src/commands/worker.rs:
```diff
@@ -32,9 +38,17 @@ impl StartWorker {
         tonic::transport::Server::builder()
             .accept_http1(true)
             .add_service(tonic_web::enable(rpc.api_service))
+            // Add the health service to the server
+            .add_service(health_service)
```
nit: I'm not sure the comment here adds much value.

bobbinth commented on bin/tx-prover/src/proxy/mod.rs:
```diff
@@ -31,6 +37,9 @@ const LOCALHOST_ADDR: &str = "127.0.0.1";
 // WORKER
 // ================================================================================================
 
+/// Worker
+///
+/// A worker is a backend server that processes requests. It is represented by its address.
 #[derive(Debug, Clone, PartialEq, Eq)]
 pub struct Worker {
```
I would write the comment as:
```
/// A worker used for processing of requests.
///
/// A worker consists of a backend service (defined by worker address) and a flag indicating
/// whether the worker is currently available to process new requests.
```

SantiagoPittella replied:
I'm updating this comment, and including the health check client info.

bobbinth commented on bin/tx-prover/src/proxy/mod.rs:
```diff
@@ -41,6 +50,17 @@ impl Worker {
+    ///
+    /// # Errors
+    /// - If the worker address cannot be converted to [WorkerConfig].
+    fn update_worker_config(workers: &[Worker]) -> Result<(), String> {
+        let worker_configs =
+            workers.iter().map(|worker| worker.try_into()).collect::<Result<Vec<_>, _>>()?;
+
+        ProxyConfig::update_workers(worker_configs)
+    }
```
Adding this method to the `Worker` struct feels a bit weird to me. Specifically, it is not very intuitive that interacting with a `Worker` would update `ProxyConfig`.

Given that this method is pretty simple and is used only once, I'd probably just get rid of it and move the code to its current callsite.

bobbinth commented on bin/tx-prover/src/proxy/mod.rs:
```diff
@@ -239,6 +254,59 @@ impl LoadBalancer {
         Some(create_workers_updated_response(session, workers_count).await)
     }
+
+    /// Create a gRPC client for the given worker address.
+    ///
+    /// It will panic if the worker URI is invalid.
+    async fn create_grpc_client(
+        &self,
+        worker_addr: &str,
+    ) -> Result<HealthClient<Channel>, tonic::transport::Error> {
+        Channel::from_shared(format!("http://{}", worker_addr))
+            .expect("Invalid worker URI")
+            .connect_timeout(self.connection_timeout_secs)
+            .timeout(self.timeout_secs)
+            .connect()
+            .await
+            .map(HealthClient::new)
+    }
```
Would it make sense to move the channel creation to the `Worker::new()` constructor? We could also store the created channel in the worker itself like so (unless it causes some complications):
```rust
pub struct Worker {
    worker: Backend,
    health_checker: HealthClient<Channel>,
    is_available: bool,
}
```

We could also move the actual health-checking logic into the worker - e.g., via something like this:
```rust
impl Worker {
    async fn is_healthy(&self) -> bool {
        ...
    }
}
```

SantiagoPittella replied:
I don't think that this approach causes any complications, and is much better than mine. I will do it that way.

bobbinth reviewed (commented):
Looks good! Thank you. I left some comments inline.

Another thing I was thinking about (not for this PR): do we actually want to maintain the list of workers in the config file? It seems like it requires quite a bit of synchronization. Maybe we remove it from there and pass the list of initial workers on proxy startup as CLI parameters?

SantiagoPittella commented on bin/tx-prover/src/commands/mod.rs:
```diff
@@ -100,6 +103,17 @@ impl ProxyConfig {
+    ///
+    /// This method will persist the new workers list to the configuration file.
+    pub(crate) fn update_workers(workers: Vec<WorkerConfig>) -> Result<(), String> {
+        let mut proxy_config = Self::load_config_from_file().unwrap();
```
Yes, in fact this should use `?`.

SantiagoPittella commented on bin/tx-prover/src/proxy/mod.rs:
```diff
@@ -514,7 +592,57 @@ impl ProxyHttp for LoadBalancer {
+                *workers = healthy_workers;
+
+                // Persist the updated worker list to the configuration file
+                if let Err(err) = Worker::update_worker_config(&workers) {
```
This might be avoid if we remove the workers from the config file.

SantiagoPittella commented:
@bobbinth @igamigo I addressed the comments and opened #1008 and #1009 , the first being related to remove the workers from the configuration file and the second to define and implement a retry policy for the health check.

tomyrd reviewed (commented):
Looks good, left a couple of questions. I want to test it locally before approving it.

bobbinth commented on bin/tx-prover/src/proxy/mod.rs:
```diff
@@ -31,15 +37,77 @@ const LOCALHOST_ADDR: &str = "127.0.0.1";
 // WORKER
 // ================================================================================================
```
This file is getting pretty big. I would probably move the `Worker` struct into a separate file - e.g., `src/proxy/worker.rs`.

SantiagoPittella replied:
Makes sense. Changing it.

bobbinth commented on bin/tx-prover/Cargo.toml:
```diff
@@ -47,13 +47,15 @@ figment = { version = "0.10", features = ["toml", "env"] }
 miden-lib = { workspace = true, default-features = false }
 miden-objects = { workspace = true, default-features = false }
 miden-tx = { workspace = true, default-features = false }
+pollster = "0.4"
```
Do we still use `pollster` somewhere? (on a quick look I didn't see where it is used).

SantiagoPittella replied:
We don't use it anymore.

bobbinth commented on bin/tx-prover/src/proxy/mod.rs:
```diff
@@ -239,6 +312,27 @@ impl LoadBalancer {
         Some(create_workers_updated_response(session, workers_count).await)
     }
+
+    /// Check the health of the workers and returns a list of healthy workers.
+    ///
+    /// Performs a health check on each worker using the gRPC health check protocol. If a worker
+    /// is not healthy, it won't be included in the list of healthy workers.
+    async fn check_workers_health(
```
One thing I'm wondering: how would this affect workers which are currently busy processing a task? If the worker is healthy, it won't have an effect. But if it isn't (i.e., the worker has crushed), wouldn't `Self::logging()` method eventually get called (presumably after some timeout) and the worker will be added back to the list of available workers?

SantiagoPittella replied:
If the worker is not healthy, `check_worker_health` will remove it from the list of workers, no matter if it is currently unavailable (i.e. is currently processing a request). And when `Self::logging` is called at the end of the process, since this particular worker is not in the list of workers anymore, it won't be added back.

bobbinth reviewed (approved):
Looks good! Thank you! I left a couple of comments inline. After these comments are addressed (as well as comments from @tomyrd and @igamigo), we can probably merge.

Though, I would follow this up with a PR to address https://github.com/0xPolygonMiden/miden-base/issues/1008 as this should simplify the structure considerably.

tomyrd reviewed (approved):
Tested it on my machine and it works correctly

SantiagoPittella commented on bin/tx-prover/src/commands/proxy.rs:
```diff
@@ -23,25 +28,33 @@ impl StartProxy {
+        server.add_service(health_check_service);
         server.add_service(lb);
-        server.run_forever();
+        tokio::task::spawn_blocking(|| server.run_forever())
+            .await
+            .map_err(|err| err.to_string())?;
```
We do so because Pingora create a new runtime in `.run_forever()` without the possibility of passing and existing one. The `tokio::task::spawn_block` was introduced in order to avoid a panic each time that we instantiate a proxy.

bobbinth commented on bin/tx-prover/src/proxy/worker.rs:
```diff
@@ -0,0 +1,102 @@
+// WORKER
+// ================================================================================================
```
nit: we usually place section separators after imports (i.e., after line 13 in this file).

bobbinth commented on bin/tx-prover/src/proxy/worker.rs:
```diff
@@ -0,0 +1,102 @@
+/// the worker is currently available to process new requests, and a gRPC health check client.
+#[derive(Debug, Clone)]
+pub struct Worker {
+    pub(crate) worker: Backend,
+    health_check_client: HealthClient<Channel>,
+    pub(crate) is_available: bool,
```
Why do some fields here need to be `pub(crate)`? We already have `Self::is_available()` method and we could probably do the same for the `worker` field (I'd probably rename it into `backend`).

bobbinth commented on bin/tx-prover/src/proxy/worker.rs:
```diff
@@ -0,0 +1,102 @@
+        }
+    }
+
+    /// Create a gRPC [HealthClient] for the given worker address.
+    ///
+    /// It will panic if the worker URI is invalid.
+    async fn create_health_check_client(
```
nit: since this does not use anything from `Self` - I'd probably make it a stand-alone helper function.

bobbinth reviewed (approved):
Looks good! Thank you! I left a few small comments inline - but we are basically good to merge.

igamigo commented on bin/tx-prover/src/commands/mod.rs:
```diff
@@ -96,10 +100,19 @@ impl ProxyConfig {
     }
+
+    /// Updates the workers in the configuration with the new list.
+    pub(crate) fn set_workers(workers: Vec<WorkerConfig>) -> Result<(), String> {
```
Maybe this could be done as part of the incoming follow-up work, but we should probably check (unless it's being dome somewhere already) that there are no duplicate workers at any point (both in the worker list and persisted config file) to avoid problems if the user accidentally adds the same address/port twice.

SantiagoPittella replied:
It is not being checked. Though the worker config is planned to be removed from the configuration file. I think we can dismiss this for now and use that issue to fix this.

igamigo commented on bin/tx-prover/src/proxy/mod.rs:
```diff
@@ -239,6 +223,27 @@ impl LoadBalancer {
+        let mut healthy_workers = Vec::new();
+
+        for worker in workers {
+            if worker.is_healthy().await {
```
Not for this PR but I think it would be nice to parallelize these checks somehow. Since they are sequential, if you have a couple of them that are failing, they can both stall all the checks by the defined timeout amount.

SantiagoPittella replied:
Maybe it will be good to add this to the #1009 issue (the one of the retries), and change a bit the scope of that issue as a more general "improve health check" in the proxy side. I'm updating the issue, let me know your opinion.

igamigo commented on bin/tx-prover/src/proxy/mod.rs:
```diff
@@ -317,6 +322,16 @@ impl RequestContext {
+/// This wrapper is used to implement the ProxyHttp trait for Arc<LoadBalancer>.
+/// This is necessary because we want to share the load balancer between the proxy server and the
+/// health check background service.
+pub struct LoadBalancerWrapper(pub Arc<LoadBalancer>);
```
nit: Because now `ProxyHttp` is being implemented for the wrapper struct, I think we could rename `LoadBalancerWrapper` into `LoadBalancer` and `LoadBalancer` could be something like `LoadBalancerState` (or maybe `WorkerState`) since the idea of adding the wrapper was to be able to easily share the state

SantiagoPittella replied:
I like `LoadBalancerState`, pretty much straightforward to wait it does. I'm renaming it.

igamigo reviewed (approved):
LGTM! Left some minor comments, but no need to tackle them in this PR

bobbinth reviewed (approved):
All looks good to me! Thank you!

@SantiagoPittella or @igamigo - feel free to merge w/e you think is appropriate.
